File: encrypt_decrypt.nim
#use gc:arc/orc for speeeeeeeed
#http://practicalcryptography.com/cryptanalysis/ for reference
import strutils
import sugar
import sequtils
import tables
import re
import math
import sets
import std/enumerate
import algorithm

const BASE = ord("a"[0])
# let alphabetSeq = collect(for num in BASE..BASE+25: $chr(num))
let alphabet = collect(for num in BASE..BASE+25: $chr(num)).join()

proc pairMin(x: seq[(float,int)]): (float,int) =
  ## The minimum value of `x`. `T` needs to have a `<` operator.
  var currMin = x[0][0]
  result = x[0]
  for i in 1..high(x):
    if x[i][0] < currMin: result = x[i]

proc loadTetragrams(): Table[string,float] = 
    var tetraTable: Table[string, float]
    for line in lines("english_quadgrams.txt"):
        let splitLine = line.split()
        let tetra = toLowerAscii(splitLine[0])
        let tetraCount  = parseFloat(splitline[1])
        tetraTable[tetra] = tetraCount
    return tetraTable

proc normaliseTetragrams(tetraTable:Table[string,float]):Table[string,float] = 
    let totalCount = sum(tetraTable.values().toSeq())
    var newTable = tetraTable
    for key in newTable.keys():
        newTable[key] = float(log10(float(newTable[key])/totalCount))
    return newTable

let tetraTable = normaliseTetragrams(loadTetragrams())

proc tetraScore(text:string):float =
    # should be <-6 for real text?
    var score:float = 0
    for i in 0..len(text)-(4): 
        if text[i..i+3] in tetraTable:
            score = score + tetraTable[text[i..i+3]]
        else:
            score = score + log10(1/(4*100))
    return score/float(len(text))

func translateText(text: string, transDict: seq[(string, string)]): string =
    return multiReplace(text, transDict)

proc filterLower(text: string, removeSpaces:bool = false): string {.inline.} =
    # filter_text = regex to remove all non a-z
    if removeSpaces:
        return re.replace(toLowerAscii(text),re"^[a-z]","")
    else:
        return re.replace(toLowerAscii(text), re"[^a-z ]", "")

proc indexCoincidence(text: string): float =
    let cleanText = re.replace(text,re" ","")
    let counter = toCountTable(cleanText)
    let denom = (len(cleanText)*(len(cleanText)+1))/26
    var iOC = 0
    for key, value in counter.pairs():
        ioc += value*(value-1)
    return float(ioc)/denom

proc caesarEncrypt(text: string, shift: int): string =
    if shift mod 26 == 0:
        return text
    let shiftedAlpha = alphabet[shift..alphabet.high] & alphabet[0..shift]
    let alphaSeq: seq[string] = alphabet.mapIt($it)
    let transDict = zip(alphaSeq, collect(for letter in shiftedAlpha: $letter))
    return translateText(text, transDict)

proc caesarDecrypt(text: string, shift: int): string =
    return caesarEncrypt(text, -shift+26)

proc absError(text: string): float =
    # real text is around 0.2, but can put below 0.5 to be safe?
    let sourceText = filterLower(readFile("holmes-gutenberg.txt"),true)
    let sourceCounter = toCountTable(sourceText)
    let textCounter = toCountTable(text)
    var error: float
    let totalChars = len(text)
    for key in textCounter.keys():
        error += abs(sourceCounter[key]/len(sourceText) - textCounter[
                key]/totalChars)
    return error

proc chiSquared(text: string): float =
    # real text is around 0.2, but can put below 0.5 to be safe?
    let sourceText = filterLower(readFile("holmes-gutenberg.txt"),true)
    let sourceCounter = toCountTable(sourceText)
    let textCounter = toCountTable(text)
    var error: float
    let totalChars = len(text)
    for key in textCounter.keys():
        error += ((sourceCounter[key]/len(sourceText) - textCounter[
                key]/totalChars)^2)/float(textCounter[
                key]/totalChars)
    return error

# # from https://stackoverflow.com/questions/8870261/how-to-split-text-without-spaces-into-list-of-words

# # Build a cost dictionary, assuming Zipf's law and cost = -math.log(probability).
# words = open("google-10000-english-no-swears.txt").read().split()
let words = readFile("google-10000-english-no-swears.txt").split()
# wordcost = dict((k, log((i+1)*log(len(words)))) for i, k in enumerate(words))
let wordcost = collect(for (i,k) in enumerate(words): (k, log10(float(i+1)*log10(float(len(words)))))).toTable()
# maxword = max(len(x) for x in words)
let maxword = max(map(words,word=>len(word)))
echo "maxword: ", maxword

# def infer_spaces(s: str):
proc inferSpaces(text:string):string = 
#     Uses dynamic programming to infer the location of spaces in a string
#     without spaces.
#     cost: list = [0]
    var cost = @[0.0]
#     # Find the best match for the i first characters, assuming cost has
#     # been built for the i-1 first characters.
#     # Returns a pair (match_cost, match_length).
#     def best_match(i):
    proc bestMatch(idx:int,cost:seq[float]): (float,int) = 
        let candidates = pairs(cost[max(0,idx-maxword)..min(idx,len(cost)-1)].reversed()).toSeq()
#         candidates = enumerate(reversed(cost[max(0, i-maxword):i]))
#       how to get min of pairs from first items???
        return pairMin(collect(for (k, c) in candidates: (float(c) + wordcost.getOrDefault($(text[idx-k-1..idx]), 1e20), (k+1)) ))
#         return min((c + wordcost.get(s[i-k-1:i], 9e999), k+1) for k, c in candidates)

#     # Build the cost array.
#     for i in range(1, len(s)+1):

    for i in 1..len(text):
        if i > 998:
            echo "cost ", cost
            echo "i ", i
        let (c,k) = bestMatch(i,cost)
#         c, k = best_match(i)
        cost.add(c)
#         cost.append(c)

#     # Backtrack to recover the minimal-cost string.
#     output = []
    var output: seq[string] = @[] 
#     i = len(s)
    var i = len(text)
#     while i > 0:
    while i > 0:
        let (c,k) = bestMatch(i,cost)
#         c, k = best_match(i)
        assert c == cost[i]
#         assert c == cost[i]
        output.add( $(text[i-k..i]) )
#         output.append(s[i-k:i])
        i = i-k
#         i -= k
    return join(output.reversed()," ")
#     return " ".join(reversed(out))


proc smartCaesarDecrypt(text: string): string =
    for shift in 0..25:
        let plaintext = caesarDecrypt(text, shift)
        if absError(plaintext) < 0.4:
            return plaintext
    echo "no solutions"

func vignereEncrypt(text: string, key: string): string =
    return collect(for (char1, char2) in zip(text, cycle(key, len(
            text)).join): 
                if char1==" "[0]: char1
                else: chr(((ord(char1)+ord(char2)-BASE-BASE) mod 26) +
                BASE)).join()

func vignereDecrypt(text: string, key: string): string =
    return collect(for (char1, char2) in zip(text, cycle(key, len(
            text)).join):
            if char1==" "[0]: char1
            else: chr((((ord(char1)-ord(char2)) + 26) mod 26) +
            BASE)).join()

proc monoSubEncrypt(text:string, key:string):string =
    let strAlphabet: seq[string] = alphabet.toSeq().mapIt( $it ) 
    var dedupedkey = toOrderedSet(filterLower(key).toSeq()).toSeq().mapIt( $it )
    for letter in strAlphabet:
        if not dedupedkey.contains(letter):
            dedupedkey.add(letter)
    let transTable = zip(dedupedkey,strAlphabet)
    return translateText(text,transTable)

proc monoSubDecrypt(text:string, key:string): string = 
    let strAlphabet: seq[string] = alphabet.toSeq().mapIt( $it ) 
    var dedupedkey = toOrderedSet(filterLower(key).toSeq()).toSeq().mapIt( $it )
    for letter in strAlphabet:
        if not dedupedkey.contains(letter):
            dedupedkey.add(letter)
    let transTable = zip(strAlphabet,dedupedkey)
    return translateText(text,transTable)


proc analyseVignere(text: string): int =
    # returns a guess of keylen using ioc ~1.7 as natural english
    for n in 1..text.high():
        if indexCoincidence(collect(for index in countup(0, text.high(),
                n): text[index]).join()) > 1.55:
            return n

iterator iterProduct(args: varargs[string], repetitions: int): string =
    # # product('ABCD', 'xy') --> Ax Ay Bx By Cx Cy Dx Dy
    # # product(range(2), repeat=3) --> 000 001 010 011 100 101 110 111
    # shamelessly stolen from python itertools
    let pools = map(sequtils.repeat(args.join(), repetitions), (n) => $n)
    var res: seq[seq[string]] = @[newSeq[string]()]
    for count, pool in enumerate(pools):
        var temp: seq[seq[string]] = @[newSeq[string]()]
        for x in res:
            for y in pool:
                if count == (repetitions-1):
                    yield (concat(x, @[$y]).join()) 
                temp.add(concat(x, @[$y]))
        res = temp

func product(args: varargs[string], repetitions: int): seq[string] =
    # # product('ABCD', 'xy') --> Ax Ay Bx By Cx Cy Dx Dy
    # # product(range(2), repeat=3) --> 000 001 010 011 100 101 110 111
    # shamelessly stolen from python itertools
    let pools = map(sequtils.repeat(args.join(), repetitions), (n) => $n)
    var res: seq[seq[string]] = @[newSeq[string]()]
    for pool in pools:
        var temp: seq[seq[string]] = @[newSeq[string]()]
        for x in res:
            for y in pool:
                temp.add(concat(x, @[$y]))
        res = temp
    return collect(for prod in res: prod.join())


proc bruteVignere(text: string, keylen: int): string =
    #TODO: add multithreading
    #~20s for 5 char key, ~2s for 4 char key , ~30s for 5char key with 800ish chars.  If result looks like gibberish, caesardecrypt it
    for possibleKey in iterProduct(alphabet, keylen-1):
        let possibleKey = "a" & possibleKey
        if len(possibleKey) == keylen and indexCoincidence(vignereDecrypt(text,
                possibleKey)) > 1.7: #can change comparison param when necessary
            return vignereDecrypt(text, possibleKey)
    echo "\nall keys exhausted"

proc smartVignere(text:string, keylen:int):string =
    var sequences: seq[string]
    var plaintext: string
    for start in 0..(keylen-1):
        sequences.add(smartCaesarDecrypt(collect(for index in countup(start,len(text)-1,keylen): text[index]).join()))
    for index in 0..len(text)-1:
        plaintext.add(sequences[index mod keylen][index div keylen])
    return plaintext

proc wordlistVignere(text: string, keylen: int): string =
    for line in lines("google-10000-english-no-swears.txt"):
        if len(line) == keylen and indexCoincidence(vignereDecrypt(text,
                line)) > 1.65: #can change comparison param when necessary
            return vignereDecrypt(text, line)
    echo "\nall keys exhausted"

assert filterLower(r"aBc.><.??.\\\\////") == "abc"

assert vignereEncrypt("attackatdawn", "lemon") == "lxfopvefrnhr"
assert vignereDecrypt("lxfopvefrnhr", "lemon") == "attackatdawn"

assert caesarEncrypt("avecaesar", 3) == "dyhfdhvdu"
assert caesarDecrypt("dyhfdhvdu", 3) == "avecaesar"


proc test() =
    let plaintext = filterLower(readFile("holmes-gutenberg.txt"),false)[0..1000]
    discard caesarEncrypt(plaintext,10)
    echo indexCoincidence(plaintext) # around 1.7
    echo tetraScore(plaintext)
    echo "plaintext:\n", plaintext[0..1000]
    let encrypted = vignereEncrypt(plaintext, "vulcans")
    assert smartVignere(encrypted,analyseVignere(encrypted)) == plaintext
    echo inferSpaces(filterLower(plaintext,true))
    assert smartCaesarDecrypt(wordlistVignere(encrypted, analyseVignere(encrypted))) == plaintext
    assert monoSubDecrypt(monoSubEncrypt("attackatdawnavecaesar","lemonbeef"),"lemonbeef") == "attackatdawnavecaesar"

when defined(test):
    test()

File: encrypt_decrypt.py
from math import log
from string import ascii_lowercase
from collections import Counter
from typing import Iterable
from math import gcd, ceil
from itertools import cycle, pairwise
from functools import lru_cache
from numpy import random  # 0.2 ms
# all funcs assume text is lowercase

BASE = ord("a")


def load_text(file: str) -> str:
    """
    Utility function which loads text file for longer sequences of text
    """
    with open(f"{file}.txt", encoding="utf8") as text:
        return text.read()


def load_wordlist() -> list[str]:
    """
    Utility function which loads text file for longer sequences of text
    """
    with open("google-10000-english-no-swears.txt") as words:
        return [word.strip() for word in words.readlines() if len(word) > 3]


words = load_wordlist()

# from https://stackoverflow.com/questions/8870261/how-to-split-text-without-spaces-into-list-of-words

# Build a cost dictionary, assuming Zipf's law and cost = -math.log(probability).
words = open("google-10000-english-no-swears.txt").read().split()
wordcost = dict((k, log((i+1)*log(len(words)))) for i, k in enumerate(words))
maxword = max(len(x) for x in words)


def infer_spaces(s: str):
    """Uses dynamic programming to infer the location of spaces in a string
    without spaces."""

    # Find the best match for the i first characters, assuming cost has
    # been built for the i-1 first characters.
    # Returns a pair (match_cost, match_length).
    def best_match(i):
        candidates = enumerate(reversed(cost[max(0, i-maxword):i]))
        return min((c + wordcost.get(s[i-k-1:i], 9e999), k+1) for k, c in candidates)

    # Build the cost array.
    cost: list = [0]
    for i in range(1, len(s)+1):
        c, k = best_match(i)
        cost.append(c)

    # Backtrack to recover the minimal-cost string.
    out = []
    i = len(s)
    while i > 0:
        c, k = best_match(i)
        assert c == cost[i]
        out.append(s[i-k:i])
        i -= k

    return " ".join(reversed(out))


def most_common(text: Iterable) -> list[tuple[str, int]]:
    # TODO: use str.count to be faster?
    return Counter([char for char in text if char not in (" ", "\n")]).most_common()


def mod_inverse(a: int) -> int:
    return pow(a, -1, 26)


def coprime(a: int, b: int) -> bool:
    return gcd(a, b) == 1


def contains_words(text, wordlist) -> float:
    count = 0
    for word in wordlist:
        if text.count(word) > 0:
            print(word)
        count += text.count(word)
    # average words per char, should be >.3 for legit text?
    return count/len(text)


def is_english(text, *cribs) -> bool:
    freq_chars = most_common(text)
    freq_onegram = most_common(
        word for word in text.split(" ") if len(word) == 1)
    if cribs != ():
        if all(crib in text for crib in cribs):
            return True
    # freq_threegram = most_common(
    #     word for word in text.split(" ") if len(word) == 3) #should be the, and for first two
    if freq_chars[0][0] == "e" and freq_chars[1][0] in ("t", "a", "i", "o") and freq_onegram[0][0] in ("a", "i"):
        return True
    return False


def normalised_common_ref():
    sample = load_text("holmes-gutenberg").lower()
    # don't filter chars in order for speeeed!
    reference = Counter(sample).most_common()
    total_chars = len([char for char in sample if char in ascii_lowercase])
    normalised_ref = dict([(char, count/total_chars)
                           for (char, count) in reference if char in ascii_lowercase])
    return normalised_ref


def pair_normalised_common_ref():
    sample = load_text("holmes-gutenberg").lower()
    filter_text = [char for char in sample if char not in ("\n")]
    reference = Counter(pairwise(filter_text)).most_common()
    total_pairs = len(filter_text)-1
    normalised_ref = dict([(pair, count/total_pairs)
                           for (pair, count) in reference])
    return normalised_ref


normalised_ref = normalised_common_ref()
normalised_ref_pairwise = pair_normalised_common_ref()


# def find_cribs() -> set[str]:
#     rare_chars = list(
#         zip(*sorted(normalised_ref.items(), key=lambda x: x[1])))[0][:5]
#     cribs = set()
#     for word in words[:]:
#         for rare_char in rare_chars:
#             if rare_char in word:
#                 cribs.add(word)
#     return cribs


# cribs = find_cribs()


def mean_abs_error(text) -> float:
    """
    Real text is < 0.01?
    """
    error = 0.0

    common_text = most_common(text)
    total_chars = len(
        [char for char in text if char in ascii_lowercase])
    normalised_text = dict([(char, count/total_chars)
                            for (char, count) in common_text if char in ascii_lowercase])
    for key in normalised_text:
        error += abs(normalised_ref[key]-normalised_text[key])
    return error/len(normalised_text)


@lru_cache
def di_mean_abs_error(text, *cribs) -> float:
    """
    Real text is < 0.01?
    """
    error = 0.0
    # filter_text = [char for char in text if char not in ("\n")]
    common_text = Counter(pairwise(text))
    total_pairs = len(text)-1
    # normalised_text = dict([(pair, count)
    #                         for (pair, count) in common_text])
    normalised_text = dict(common_text)  # remove dict to save a few ms
    for key in normalised_text:
        try:
            error += abs((normalised_ref_pairwise[key]
                         * total_pairs)-normalised_text[key])
        except KeyError:
            # print("Error! missing key, key skipped")
            error += normalised_text[key]
    for crib in cribs:  # if using more than 3 cribs, prob change the error mult
        if crib in text:
            error *= 0.8  # some value that can be finetuned
    return ((error/total_pairs))


def mean_square_error(text) -> float:
    """
    Real text is < 1e-4?
    """
    error = 0.0

    common_text = most_common(text)
    total_chars = len(
        [char for char in text if char in ascii_lowercase])
    normalised_text = dict([(char, count/total_chars)
                            for (char, count) in common_text if char in ascii_lowercase])
    for key in normalised_text:
        error += (normalised_ref[key]-normalised_text[key])**2
    return error/len(normalised_text)


def index_coincidence(text: str) -> float:
    """
    For a given ciphered message, the value for the IoC allows us to filter the list of ciphering methods

    If the Index of coincidence is high (close to 0.070)
    i.e. similar to plain text, then the message has probably been crypted using
    a transposition cipher (letters were shuffled) or a monoalphabetic substitution

    If the Index of coincidence is low (close to 0.0385)
    i.e. similar to a random text, then the message has probably been crypted using a polyalphabetic cipher
    (a letter can be replaced by multiple other ones).

    The lower the coincidence count, the more alphabets have been used.
    Example: Vigenere cipher with a key of length 4 to 8 letters have an IC of about 0.045±0.05
    """
    coincidence = denominator = 0
    common_chars = Counter(text).most_common()
    denominator = len(text)
    if denominator == 0:
        return 0
    for _, freq in common_chars:
        coincidence += freq*(freq-1)
    return coincidence/(denominator*(denominator-1))


def caesar_encrypt(text: str, shift: int) -> str:

    translate_table = {ordinal: (((ordinal +
                       shift)-BASE) % 26)+BASE for ordinal in range(BASE, BASE+26)}
    return text.translate(translate_table)


def caesar_decrypt(text: str, shift: int) -> str:
    return caesar_encrypt(text, -shift)


def smart_caesar_decrypt(text: str) -> str | None:
    for shift in range(26):
        plaintext = caesar_decrypt(text, shift)
        if mean_abs_error(plaintext) < 0.025:
            return plaintext
        print(mean_abs_error(plaintext))
    return None


def rotate(box: list[str], width: int, height: int) -> list[str]:
    return [box[y][x] for x in range(width) for y in range(height)]


def box_caesar_encrypt(text: str, width: int, sep: str = "~") -> str:
    height = ceil(len(text)/width)
    box = [text[subsection:subsection+width]
           for subsection in range(0, len(text), width)]
    if (len(text) % width) != 0:
        box[-1] += (width-(len(text) % width))*sep
    return "".join(row for row in rotate(box, width, height))


def box_caesar_decrypt(text: str, width: int) -> str:
    height = ceil(len(text)/width)
    box = [[""]*width for y in range(height)]
    for id, char in enumerate(text):
        x, y = divmod(id, height)
        box[y][x] = char
    return "".join("".join(row) for row in box).replace("~", "")


def transposition(text: str, width: int, indices: list[int]) -> str | None:
    box = [text[subsection:subsection+width]
           for subsection in range(0, len(text), width)]
    height = ceil(len(text)/width)
    new_box = [[""]*width for _ in range(height)]
    for y in range(height):
        for x, id in enumerate(indices):
            new_box[y][id] = box[y][x]
    return "".join(["".join(row) for row in new_box])


def affine_encrypt(text: str, a: int, b: int) -> str:
    translate_table = {}
    for ord_digit in range(26):
        translate_table[ord_digit+BASE] = ((a*ord_digit + b) % 26) + BASE

    return text.translate(translate_table)


def affine_decrypt(text: str, a: int, b: int) -> str:
    a = mod_inverse(a)
    b = a*-b
    return affine_encrypt(text, a, b)


def brute_affine_decrypt(ciphertext: str) -> str | None:
    for a in range(1, 12):
        for b in range(26):
            if coprime(a, 26):
                plaintext = affine_decrypt(ciphertext, a, b)
                if mean_abs_error(plaintext) < 0.01:
                    return plaintext
    return None


def vignere_encrypt(text, key):
    filter_text = "".join([char for char in text if char in ascii_lowercase])
    return "".join([chr(((ord(char_1)+ord(char_2)-BASE-BASE) %
                         26) + BASE) for (char_1, char_2) in
                    zip(filter_text, cycle(key))])
    # oneliner, hell yeah!! fuck you future me!! (assumes both are lowercase, ur fault if you don't preprocess)
    # https://en.wikipedia.org/wiki/Vigen%C3%A8re_cipher#Algebraic_description


def vignere_decrypt(text, key):
    return "".join([chr(((ord(char_1)-ord(char_2)) %
                         26) + BASE) for (char_1, char_2) in zip(text, cycle(key)) if char_1 in ascii_lowercase])


def smart_vignere(text: str, keylen: int):
    """
    Ignore for now
    use https://www.cipherchallenge.org/wp-content/uploads/2020/12/Five-ways-to-crack-a-Vigenere-cipher.pdf
    """
    parts = [text[start::keylen].ljust((len(text)//keylen)+1)
             for start in range(keylen)]
    return "".join([part for item in zip(*((part) for part in parts)) for part in item])


def ultra_smart_substitution_decrypt(text: str, *cribs: str, is_deterministic: bool = False, rand_seed=64):
    # for consistency
    if is_deterministic:
        random.seed(rand_seed)
    # based on https://sci-hub.se/10.1080/0161-119591883944
    count = 0
    for char in ascii_lowercase:
        if char not in text:
            text += char  # hack to prevent trans table missing letters
            count += 1
    error = 1.0
    revert = 0
    text_freqs = most_common(char for char in text if char in ascii_lowercase)
    if count > 0:
        text = text[:-count]
    initial_guess = zip(
        (x[0] for x in normalised_common_ref()), (x[0] for x in text_freqs))
    # placeholder for something smarter
    # weights = list(1/26 for _ in range(26))
    guess = "".maketrans(dict(initial_guess))  # type: ignore
    error = di_mean_abs_error(text.translate(guess), *cribs)
    guess_keys = list(guess.keys())
    # TODO: use combos
    x = 0
    if is_deterministic:
        pass
    else:
        while True:
            x += 1
            # choose two random letters, could be weighted
            # if we want to never repeat, shuf combos, removing when used until revert
            # when we revert, recreate combos, reshuffled?
            # choice_1, choice_2 = n_combos[-1]
            # remove last 2 elements
            choice_1, choice_2 = random.choice(
                guess_keys, 2, False)
            # swap them
            guess[choice_1], guess[choice_2] = guess[choice_2], guess[choice_1]
            new_error = di_mean_abs_error(text.translate(guess), *cribs)
            if new_error >= error:
                revert += 1
                # undo swap
                guess[choice_1], guess[choice_2] = guess[choice_2], guess[choice_1]
            else:
                error = new_error
                revert = 0
                # n_combos = combos.copy()
            if revert > 10000:  # tune this for speed against accuracy
                print("cycles = ", x, "\n error = ", error, "\n", guess)
                return text.translate(guess)


def test():
    assert ("hello there" == (caesar_decrypt(
        caesar_encrypt("hello there", 6), 6)))
    assert ("helloOO there" == affine_decrypt(
        affine_encrypt("helloOO there", 3, 7), 3, 7))
    assert (box_caesar_encrypt("AVECAESAR", 3) == "ACSVAAEER")
    assert (box_caesar_encrypt("AVECAESAR", 4) == "AARVE~ES~CA~")
    assert (box_caesar_decrypt("ACSVAAEER", 3) == "AVECAESAR")
    assert (box_caesar_decrypt("AARVE~ES~CA~", 4) == "AVECAESAR")
    assert (vignere_decrypt(vignere_encrypt(
        "attackatdawn", "lemon"), "lemon") == "attackatdawn")


def multi():
    from multiprocessing import Pool
    text = ""
    with Pool(8) as p:
        p.map(ultra_smart_substitution_decrypt, (text, "jade"))


def main():
    text = load_text(
        "ciphertexts/challenge-4-b").lower().replace("\n", " ").replace(" ", "")
    decrypt = ultra_smart_substitution_decrypt(text, "unable")
    if decrypt is None:
        raise TypeError
    print(infer_spaces(decrypt), mean_abs_error(decrypt))
    # for crib in cribs:
    #     if crib in decrypt:
    #         print(crib)

    # ciphertext = load_text("ciphertext").lower()
    # print(brute_affine_decrypt(ciphertext))


test()
if __name__ == "__main__":
    main()
